tests/e2e/kafkatopic_webhook.go

```diff
@@ -0,0 +1,318 @@
+// Copyright © 2023 Cisco Systems, Inc. and/or its affiliates
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"strings"
+	"text/template"
+
+	"github.com/Masterminds/sprig"
+	"github.com/gruntwork-io/terratest/modules/k8s"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+// TODO(mihalexa): move to k8s.go
+const (
+	dryRunStrategyNone   string = "none"
+	dryRunStrategyClient string = "client"
+	dryRunStrategyServer string = "server"
+)
+
+// TODO(mihalexa): move to k8s.go
+// applyK8sResourceFromTemplateWithDryRun is copy of applyK8sResourceFromTemplate which calls a "--dry-run=<strategy>" kubectl command
+func applyK8sResourceFromTemplateWithDryRun(kubectlOptions k8s.KubectlOptions, templateFile string, values map[string]interface{}, dryRunStrategy string) (string, error) {
+	By(fmt.Sprintf("Generating K8s manifest from template %s for dry-run apply", templateFile))
+	var manifest bytes.Buffer
+	rawTemplate, err := os.ReadFile(templateFile)
+	if err != nil {
+		return "", err
+	}
+	t := template.Must(template.New("template").Funcs(sprig.TxtFuncMap()).Parse(string(rawTemplate)))
+	err = t.Execute(&manifest, values)
+	if err != nil {
+		return "", err
+	}
+
+	By("Replicating terratest's k8s.KubectlApplyFromStringE")
+	tmpfile, err := k8s.StoreConfigToTempFileE(GinkgoT(), manifest.String())
+	if err != nil {
+		return "", err
+	}
+	defer os.Remove(tmpfile)
+
+	return k8s.RunKubectlAndGetOutputE(GinkgoT(), &kubectlOptions, "apply", "-f", tmpfile, "--dry-run="+dryRunStrategy, "--output=yaml")
+}
+
+func testWebhooks() bool {
+	return When("Testing webhooks", func() {
+		// temporary section; to be refactored after kubeconfig injection PR
+		var kubectlOptions k8s.KubectlOptions
+		var err error
+		kubectlOptions, err = kubectlOptionsForCurrentContext()
+		if err != nil {
+			GinkgoT().Fail()
+		}
+
+		testWebhookKafkTopic(kubectlOptions)
+	})
+}
+
+func testWebhookKafkTopic(kubectlOptions k8s.KubectlOptions) {
+	kubectlOptions.Namespace = koperatorLocalHelmDescriptor.Namespace
+	requireCreatingKafkaCluster(kubectlOptions, "../../config/samples/simplekafkacluster.yaml")
+	testWebhookCreateKafkaTopic(kubectlOptions)
```

Comment:
In the current implementation of e2e tests shouldn't this be done by this point? Not that these should cause problems in that case, it just might make stuff slower.

Reply:
I am not 100% sure about what you meant in the question. I will try to answer what I understood from it but please follow up if I didn't get it.

I favored having the webhook testing not rely on other previous tests and the flexibility this gives us.

There a few things I considered : 
1. our resources have different prerequisites (e.g. KafkaTopic requires that the indicated KafkaCluster exists) and we need a clear place to aggregate those
    * this PR also aimed to provide a structure where other webhook e2e tests (and their prerequisites) could easily slot in ;   
        * e.g. kafkatopics may only require a/any kafkacluster but other things may require more intricate or specific setups 
2. our KafkaCluster related testing may not always be done on the absolutely most minimal setting possible while for the purposes of KafkaTopic webhook testing we don't care about the kafkacluster resource (so minimum is best) but there is no way to go around not having the KafkaCluster.
    * this structure allowed me to decouple what kafkaclusters we test when we test kafkaclusters from what kafkaclusters we use when we test webhooks

Reply:
Okay, I'm fine with this reasoning, though I'd like to see how this looks when [Marton's PR](https://github.com/banzaicloud/koperator/pull/1020) is merged.

Reply:
Yeah I think either we have to do everything from deploying components here or only do the webhooks and let the rest be built up at somewhere else, we are in the middle here which is not good because ensuring the kafka cluster is not the webhook test's responsibility.

Reply:
We can create an explicit or implicit dependency on the KafkaCluster CR name, remove the creation and cleanup steps and reorder the function calls in the suite file so that webhook testing comes right after one of the KafkaCluster creations we do.

One worry I voiced is about how such code scales for other webhooks and their CRs and their prerequisites and dependencies.   
For example, explicit dependency injection (via function parameters like a KafkaClusterName) might mean we would keeeeep adding params to function definitions and that's not something we want either. 
Or it could be "implicit" and we just "assume" certain constants are used for names - which we already do "here and there".

I'll explain my initial stance: I wanted to separate the setup steps for each webhook because I worried the prerequisites between webhooks could vary enough to make another type of implementation very messy. That's why the sequence of I am/was proposing was : testWebhooks() -> testWebhookKafkaTopic(kubectlOptions) -> <funcs for dependencies and setup creation specific to kafkatopic> -> testCreate + testUpdate -> teardown.  

I can be convinced to, for instance, not have the setup phase as part of `testWebhookKafkaTopic()`.

Reply:
> We can create an explicit or implicit dependency on the KafkaCluster CR name, remove the creation and cleanup steps and reorder the function calls in the suite file so that webhook testing comes right after one of the KafkaCluster creations we do.

Explicit rather, but otherwise this was my thought, yes.

> I can be convinced to, for instance, not have the setup phase as part of testWebhookKafkaTopic().

Yeah with this naming IMO the setup/teardown of the components or clusters shouldn't be part of the function. Either we should name it `testKafkaClusterCreateAndWebhookKafkaTopic()` or we need to make something else do the setup.